google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -0,0 +1,48 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/pubsub/internal/tracing_message_batch.h"
+#include "google/cloud/pubsub/internal/message_batch.h"
+#include "google/cloud/pubsub/version.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "opentelemetry/context/runtime_context.h"
+#include "opentelemetry/trace/context.h"
+#include "opentelemetry/trace/span.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+void TracingMessageBatch::SaveMessage(pubsub::Message m) {
+  auto active_span = opentelemetry::trace::GetSpan(
+      opentelemetry::context::RuntimeContext::GetCurrent());
+  message_root_spans_.push_back(active_span);
+  child_->SaveMessage(std::move(m));
+}
+
+void TracingMessageBatch::Flush() { child_->Flush(); }
+
+void TracingMessageBatch::FlushCallback() { child_->FlushCallback(); }
```

Comment:
When is `message_root_spans_` cleared?  I would expect one of these functions to do so?